### PR TITLE
Fix topic operator loop for unmanaged topics

### LINKED ISSUE #10380: [Bug]: Infinite topic sync in topic operator
### Bug Description

When the specs on an unmanaged KafkaTopic resource (annotated with `strimzi.io/managed="false"`) is different from the topic on the cluster, the topic operator enters a loop and continuously tries updates the resourceVersion.

Seems like it's trying to consolidate the state, but as is unmanaged it can't and continues in the loop forever.

### Steps to reproduce

You can do the following script:

```
helm install strimzi-cluster-operator --set replicas=1 --version 0.42.0 oci://quay.io/strimzi-helm/strimzi-kafka-operator

# Create Kafka cluster:
kubectl apply -f - <<EOF 
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-kafka
spec:
  kafka:
    replicas: 1
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
    storage:
      type: persistent-claim
      size: 1Gi
  zookeeper:
    replicas: 1
    storage:
      type: persistent-claim
      size: 1Gi

  entityOperator:
    topicOperator: {}
EOF
kubectl wait --for=condition=ready --timeout=10m kafka/my-kafka

# Create a topic
kubectl apply -f - <<EOF
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: sync-test
  labels:
    strimzi.io/cluster: my-kafka
spec:
  replicas: 1
  partitions: 1
  config:
    retention.ms: 432000000
EOF

sleep 5

# Annotate topic as unmanaged
kubectl annotate kafkatopic sync-test strimzi.io/managed="false"

# Path the topic with updated retention
kubectl patch kafkatopic sync-test --type='merge' -p '{"spec": {"config": {"retention.ms": "432000001"}} }'
```

### Expected behavior

Topic operator should entirely ignore the KafkaTopic resources if it is set to unmanaged

### Strimzi version

0.42.0

### Kubernetes version

Kubernetes 1.30.0

### Installation method

helm chart

### Infrastructure

Kind, also confirmed with EKS

### Configuration files and logs

all relevant stuff is already in the steps to reproduce

### Additional context

I would like to work on this issue :)

### PR DESCRIPTION
**Pull request by padilo**
### Type of change

- Bugfix

### Description

Fixes a bug in the topic operator with unmanaged topics, that keeps updating the custom resources even isn't going to update the Kafka Topic

Closes #10380

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: topic-operator/src/main/java/io/strimzi/operator/topic/BatchingTopicController.java [hunk 1]
```diff
@@ -1195,21 +1206,20 @@ private void updateStatusForException(ReconcilableTopic reconcilableTopic, Excep
...
-        if (!TopicOperatorUtil.isManaged(reconcilableTopic.kt())) {
-            return;
-        }
         var oldStatus = Crds.topicOperation(kubeClient)
             .inNamespace(reconcilableTopic.kt().getMetadata().getNamespace())
             .withName(reconcilableTopic.kt().getMetadata().getName()).get().getStatus();
-        if (statusChanged(reconcilableTopic.kt(), oldStatus)) {
+
+        StatusDiff statusDiff = new StatusDiff(oldStatus, reconcilableTopic.kt().getStatus());
+        if (!statusDiff.isEmpty() || differentConditions(oldStatus.getConditions(), reconcilableTopic.kt().getStatus().getConditions())) {
```
Comment (scholzj) [inline topic-operator/src/main/java/io/strimzi/operator/topic/BatchingTopicController.java:1214]: The StatusDiff should include the conditions. So why do we need the second part of the condition?

**Review by scholzj — COMMENTED** [review]
This looks like the good direction. I left one comment. But the tests also seemed to have failed in the Topic Operator which seems to be related I guess?

File: topic-operator/src/main/java/io/strimzi/operator/topic/BatchingTopicController.java (hunk 1 above)
Comment (padilo) [reply topic-operator/src/main/java/io/strimzi/operator/topic/BatchingTopicController.java:1214]: You are right, we can remove it, I misslooked how StatusDiff evaluates the conditions path.

I'll remove it and have a look at the tests as well.

**Review by fvaleri — COMMENTED** [review]
Hi, thanks for working on this. I like the StatusDiff logic and the new Unmanaged condition type, but they require some additional changes.

The `setObservedGeneration` and `setTopicName` should be computed before the StatusDiff to cover the corner case of config changes when KT is ready (there is no status update). This is why the following integration test is failing:

- `shouldUpdateTopicInKafkaWhenListConfigChangedInKube`

You should also handle the corner case of "new AND unmanaged" KTs been created (the expected status should be "Unmanaged" instead of "Ready"). This is why the following integration tests are failing:

- `shouldRestoreFinalizerIfRemoved` (unmanaged KTs only)
- `shouldNotCreateTopicInKafkaWhenUnmanagedTopicCreatedInKube`

> Likely it should not have the topic ID set either as it will not know if it changed or not?

I agree that we should reset the topicId when the KT is paused/unmanaged, and set the topicId when switching back to unpaused/managed. This is actually a separate bug, so I logged #10467 that we should address with a dedicated PR.

**Comment by fvaleri** [comment]
@padilo I opened a new PR to address the issue with unmanaged topics that are partially reconciled to add finalizers. This should make it easier to complete this bugfix.

**Comment by padilo** [comment]
Tests should be fixed, but I still need to have a look at the test and remove awaitability lib :)

Sorry for the delay on this.

**Comment by fvaleri** [comment]
> Sorry for the delay on this.

No problem.

Can you please rebase and fix checkstyle errors?

**Comment by padilo** [comment]
I think this is already done in a different PR, so maybe this should be closed already :)

**Comment by fvaleri** [comment]
> I think this is already done in a different PR, so maybe this should be closed already :)

If you are referring to #10481, that addresses a different issue. There is some overlap (introduction of the "Unmanaged" status condition) but there is no fix for the tight reconciliation loop reported by you. The important bit to deliver here is the introduction of StatusDiff.

**Review by scholzj — COMMENTED** [review]
I agree with Federico. This still fixes the main issue.

**Review by fvaleri — COMMENTED** [review]
@padilo I tested your changes after removing `synchronizableAndDifferentGenerations` and they look good. If you remove that code, then we are good to go. Thanks.

**Comment by padilo** [comment]
done :)

**Comment by fvaleri** [comment]
> done :)

There is an unused import to fix.

**Review by scholzj — APPROVED** [review]
LGTM assuming the system tests pass. Thanks for the PR.

**Review by fvaleri — APPROVED** [review]
Thanks.
